### ISSUE TITLE
Fine-tuning with Multiple LoRAs

### ISSUE THREAD

**Issue by xymou** [huggingface/peft#2436]
Thanks for your valuable work!

I would like to know if it's possible to jointly train two LoRAs while only loading one base model. The overall output depends on the respective outputs of LoRA1 and LoRA2. For example, logits1 is obtained from the base model with LoRA1, and logits2 is obtained from the base model with LoRA2. I have tried the following code

```python
model.add_adapter(lora_1)
model.add_adapter(lora_2)
model.enable_adapters()

model.set_adapter("lora_1")
logits1 = model(input_ids).logits # use model with lora1 to get output
model.set_adapter("lora_2")
logits2 = model(input_ids).logits # use model with lora2 to get output
logits = logits1+logits2
loss=loss_fct(logits, labels)
loss.backward()
```

but it seems there might be some issues:
1. Once set_adapter(lora2) is called, LoRA1 no longer receives gradients; 
2. If I modify the source code of set_adapter to make both requires_grad=True, would that be correct? 

What I'm confused about is, after I execute set_adapter(lora2), does the model perform computations using the base model with LoRA2 (as I hope), or does it use the base model with both LoRA1 and LoRA2 combined?

I'm looking forward to your help! Thank you!

**Comment by BenjaminBossan** [comment]
> 1. Once set_adapter(lora2) is called, LoRA1 no longer receives gradients;

What exactly do you mean? That the existing gradients are erased? That should not happen. But in general, if adapter 2 is active, adapter 1 should not get any gradients, since it doesn't participate in the forward pass.

> 2\. If I modify the source code of set_adapter to make both requires_grad=True, would that be correct?

If you want to have multiple active adapters at the same time, you can activate them like this:

```python
model.base_model.set_adapter(["lora_1", "lora_2"])
```

Also one thing that can easily go wrong: When you create the optimizer and pass the parameters, make sure that it includes the parameters for both adapters, otherwise only one gets updated.

**Comment by xymou** [comment]
> > 1. Once set_adapter(lora2) is called, LoRA1 no longer receives gradients;
> 
> What exactly do you mean? That the existing gradients are erased? That should not happen. But in general, if adapter 2 is active, adapter 1 should not get any gradients, since it doesn't participate in the forward pass.
> 
> > 2. If I modify the source code of set_adapter to make both requires_grad=True, would that be correct?
> 
> If you want to have multiple active adapters at the same time, you can activate them like this:
> 
> model.base_model.set_adapter(["lora_1", "lora_2"])
> Also one thing that can easily go wrong: When you create the optimizer and pass the parameters, make sure that it includes the parameters for both adapters, otherwise only one gets updated.


Thank you for your timely reply! In my case, adapter 1 and adapter 2 are both supposed to participate in the forward pass. I hope to get the result of adapter 1 first, and then get the result of adapter 2, they work separately but jointly contribute to the final logits.
If I set
> model.base_model.set_adapter(["lora_1", "lora_2"])

adapter 1 and adapter 2 will work simultaneously, rather than working sequentially, right? I would like to know how to achieve the latter, where both adapters participate in the forward propagation, but they are not activated simultaneously. In other words, I would like to know if it is possible to implement the following logic under the premise of **loading only one base model?**
```python
base_model1 = AutoModelForCausalLM.from_pretrained(...)
base_model1.set_adapter(lora1)
base_model2 = AutoModelForCausalLM.from_pretrained(...) # the same as base_model1
base_model2.set_adapter(lora2)

logits1 = base_model1(input_ids).logits
logits2 = base_model2(input_ids).logits
loss = loss_fct(logits1+logits2, labels)
```

**Comment by BenjaminBossan** [comment]
Yes, what I mentioned about setting both adapters would be if they are participating at the same time in the forward pass. If you want to use them sequentially, then just calling `set_adapter` with one at a time is correct.

Here is roughly the code I would use to achieve what you described:

```python
from peft import LoraConfig, get_peft_model

base_model = AutoModelForCausalLM.from_pretrained(...)
config1 = LoraConfig(...)
config2 = LoraConfig(...)

model = get_peft_model(base_model, config1, adapter_name="adapter1")
model.add_adapter(adapter_name="adapter2", peft_config=config2)
# now both adapters exist on the model, adaper1 is still active

# ensure that ALL relevant parameters are known to the optimizer
optimizer = torch.optim.AdamW(model.parameters(), ...)

...
# training loop
model.set_adapter("adapter1")
logits1 = model(input_ids).logits
model.set_adapter("adapter2")
logits2 = model(input_ids).logits
loss = loss_fct(logits1+logits2, labels)
```

**Comment by xymou** [comment]
> Yes, what I mentioned about setting both adapters would be if they are participating at the same time in the forward pass. If you want to use them sequentially, then just calling `set_adapter` with one at a time is correct.
> 
> Here is roughly the code I would use to achieve what you described:
> ```python
> from peft import LoraConfig, get_peft_model
> 
> base_model = AutoModelForCausalLM.from_pretrained(...)
> config1 = LoraConfig(...)
> config2 = LoraConfig(...)
> 
> model = get_peft_model(base_model, config1, adapter_name="adapter1")
> model.add_adapter(adapter_name="adapter2", peft_config=config2)
> # now both adapters exist on the model, adaper1 is still active
> 
> # ensure that ALL relevant parameters are known to the optimizer
> optimizer = torch.optim.AdamW(model.parameters(), ...)
> 
> ...
> # training loop
> model.set_adapter("adapter1")
> logits1 = model(input_ids).logits
> model.set_adapter("adapter2")
> logits2 = model(input_ids).logits
> loss = loss_fct(logits1+logits2, labels)
>```


Thank you for your patient reply! I initially used code like this, but during the intermediate checks, I found that only lora2 had gradients.

```python
optimizer = torch.optim.AdamW(model.parameters(), ...)
...
# training loop
model.set_adapter("adapter1")
logits1 = model(input_ids).logits
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter1.weight.grad)   None
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter1.weight.requires_grad)   True
model.set_adapter("adapter2")
logits2 = model(input_ids).logits
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter2.weight.grad)   None
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter2.weight.requires_grad)   True
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter1.weight.requires_grad)   False
loss = loss_fct(logits1+logits2, labels)
loss.backward()
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter1.weight.grad)   None
# print(model.model.model.layers[0].self_attn.q_proj.lora_A.adapter2.weight.grad)   Not None
```
only the later adapter (lora2) successfully receive gradients.

**Comment by BenjaminBossan** [comment]
Could you try enabling both adapters before computing the loss? I.e.:

```python
...
model.base_model.set_adapter(["adapter1", "adapter2"])
loss = loss_fct(logits1+logits2, labels)
loss.backward()
```

**Comment by xymou** [comment]
@BenjaminBossan Thank you for your help! I resolved this issue by instantiating two base models and having them share parameters.

**Comment by LouisDo2108** [comment]
@xymou could you please share how you “instantiated two base models and had them share parameters.”. Did you further modify the training loop inside Huggingface's Trainer?

Thank you.
